server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Keywords.scala

```diff
@@ -0,0 +1,151 @@
+      result
+    }
+
+    foo(value)
+  }
+}
+
+trait KeywordsValidator {
+      def validate(keywords: Set[Keyword]): Boolean
+}
+
+object KeywordsValidator {
+      val THROW_ON_IMAP_NON_EXPOSED_KEYWORDS: KeywordsValidator = (keywords: Set[Keyword]) => {
+        if (keywords.exists(KeywordFilter.FILTER_IMAP_NON_EXPOSED_KEYWORDS)) {
+          throw new IllegalArgumentException("Does not allow to update 'Deleted' or 'Recent' flag")
+        }
+        true
+      }
+
+      val IGNORE_NON_EXPOSED_IMAP_KEYWORDS: KeywordsValidator = (keywords: Set[Keyword]) => {
+        def foo(keywords: Set[Keyword]): Option[Keyword] = ???
+        foo(keywords).isDefined
+      }
+    }
+
+object KeywordFilter {
+      val FILTER_IMAP_NON_EXPOSED_KEYWORDS: Keyword => Boolean = (keyword: Keyword) => keyword.isExposedImapKeyword()
+      val KEEP_ALL: Keyword => Boolean = (_: Keyword) => true
+}
+
+object KeywordsFactory {
+    val STRICT_KEYWORDS_FACTORY = new KeywordsFactory(KeywordsValidator.THROW_ON_IMAP_NON_EXPOSED_KEYWORDS, KeywordFilter.KEEP_ALL, ToKeyword.STRICT)
+    val LENIENT_KEYWORDS_FACTORY = new KeywordsFactory(KeywordsValidator.IGNORE_NON_EXPOSED_IMAP_KEYWORDS, KeywordFilter.FILTER_IMAP_NON_EXPOSED_KEYWORDS, ToKeyword.LENIENT)
+
+    private[model] trait KeywordFilter extends IterableOnce[Keyword] {}
+}
+
+object Keywords {
+  val DEFAULT_VALUE: ImmutableSet[Keywords] = ImmutableSet.of()
+}
+
+case class Keywords(keywords: Set[Keyword]) {
+
+  def combiner(firstBuilder: FlagsBuilder, secondBuilder: FlagsBuilder): FlagsBuilder = firstBuilder.add(secondBuilder.build)
+  def accumulator(accumulator: FlagsBuilder, keyword: Keyword): FlagsBuilder = accumulator.add(keyword.asFlags())
+
+  /**
+   * This Keywords factory will filter out invalid keywords (containing inavlid character, being too long or being non
+   * exposed IMAP flags)
+   */
+  def strictFactory: KeywordsFactory = STRICT_KEYWORDS_FACTORY
+
+  /**
+   * This Keywords factory will throw upon invalid keywords (containing inavlid character, being too long or being non
+   * exposed IMAP flags)
+   *
+   * @throws IllegalArgumentException
+   */
+  def lenientFactory: KeywordsFactory = LENIENT_KEYWORDS_FACTORY
+
+  def asFlags(): Flags = keywords.asJava.stream().reduce(FlagsBuilder.builder, this.accumulator, this.combiner).build
+
+  def asFlagsWithRecentAndDeletedFrom(originFlags: Flags): Flags = {
+    val flags = asFlags
+    if (originFlags.contains(Flags.Flag.DELETED)) flags.add(Flags.Flag.DELETED)
+    if (originFlags.contains(Flags.Flag.RECENT)) flags.add(Flags.Flag.RECENT)
+    flags
+  }
+
+  def asMap: ImmutableMap[String, Boolean] = keywords.asJava.stream()
+    .collect(Guavate.toImmutableMap((keyword: Keyword) => keyword.getFlagName(), (_: Keyword) => Keyword.FLAG_VALUE))
+
+  def getKeywords: Set[Keyword] = keywords
+
+  def contains(keyword: Keyword): Boolean = keywords.contains(keyword)
+
+  override def equals(other: Any): Boolean =  other match {
```

Comment:
Overiding equals, hashCode & toString is not necessary if you have a case class.